### PR TITLE
Add Symbolic Quantum Mechanics

### PR DESCRIPTION
**Pull request by certik**
This adds the Symbolic Quantum Mechanics section.

@ellisonbg I took the text you sent me some time ago and polished it and integrated it in the article. Would you mind taking it from here with your students?

### CODE REVIEW

File: quantum_mechanics.tex [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+\verb|sympy.physics.quantum| subpackage. At the base level, this subpackage has
+Python objects to represent the different mathematical objects relevant in
+quantum theory \cite{sakurai}: states (bras and kets), operators (unitary,
+hermitian, etc.) and basis sets as well as operations on these objects such as
+tensor products, inner products, outer products, commutators, anticommutators,
+etc.  The base objects are designed in the most general way possible to enable
+any particular quantum system to be implemented by subclassing the base
+operators to provide system specific logic.
+
+The `quantum` subpackage has a general purpose `qapply` function that is
```
Comment (asmeurer) [inline quantum_mechanics.tex]: Assumedly the ` here should be \verb

Comment (certik) [inline quantum_mechanics.tex]: Good catch, fixed in 60e3df4bce35c752aab0f4166701f6302429e29b.

File: quantum_mechanics.tex [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+tensor products, inner products, outer products, commutators, anticommutators,
+etc.  The base objects are designed in the most general way possible to enable
+any particular quantum system to be implemented by subclassing the base
+operators to provide system specific logic.
+
+The \verb|quantum| subpackage has a general purpose \verb|qapply| function that
+is capable of applying operators to states symbolically as well as simplifying
+a wide range of symbolic expressions involving different types of products and
+commutator/anticommutators. The state and operator objects also have a rich API
+for declaring their representation in a particular basis. This includes the
```
Comment (flacjacket) [inline quantum_mechanics.tex]: particular basis and transforming between bases.

File: quantum_mechanics.tex [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+On top of this base set of objects, a number of specific quantum systems have
+been implemented. First, we have implemented the traditional algebra for
+quantum angular momentum \cite{zare}. This allows the different spin operators
+($S_x$, $S_y$, $S_z$) and their eigenstates to be represented in any basis and
+for any spin quantum number. Facilities for Clebsch-Gordan Coefficients, Wigner
+Coefficients, rotations, and angular momentum coupling are also present in
+their symbolic and numerical forms.
+
+Second we have implemented a full set of states and operators for symbolic
```
Comment (flacjacket) [inline quantum_mechanics.tex]: "Second,"

File: quantum_mechanics.tex [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+a wide range of symbolic expressions involving different types of products and
+commutator/anticommutators. The state and operator objects also have a rich API
+for declaring their representation in a particular basis. This includes the
+ability to specify a basis for a multidimensional system using a complete set
+of commuting Hermitian operators.
+
+On top of this base set of objects, a number of specific quantum systems have
+been implemented. First, we have implemented the traditional algebra for
+quantum angular momentum \cite{zare}. This allows the different spin operators
+($S_x$, $S_y$, $S_z$) and their eigenstates to be represented in any basis and
```
Comment (flacjacket) [inline quantum_mechanics.tex]: We call these `J` rather than `S`, also add `J^2`

File: quantum_mechanics.tex [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+for any spin quantum number. Facilities for Clebsch-Gordan Coefficients, Wigner
```
Comment (flacjacket) [inline quantum_mechanics.tex]: Maybe expand on this, replace this sentence with:
A rotation operator, which may be defined symbolically or numerically, is also implemented to rotate spin eigenstates. Functionality for coupling and uncoupilng of arbitrary spin states is provided, including symbolic representations of Clebsch-Gordon coefficients and Wigner symbols.

File: quantum_mechanics.tex [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+Coefficients, rotations, and angular momentum coupling are also present in
+their symbolic and numerical forms.
+
+Second we have implemented a full set of states and operators for symbolic
+quantum computing \cite{nielsen} Multidimensional qubit states can be
```
Comment (flacjacket) [inline quantum_mechanics.tex]: Period after citation

File: quantum_mechanics.tex [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+etc.) and two qubit ($CNOT$, etc.) gates (unitary operators) are provided.
+These can be represented as matricies (sparse or dense) or made to act on
+qubits symbolically without representation. With these gates, it is possible to
+implement a number of basic quantum circuits including the quantum Fourier
+transform, quantum error correction, quantum teleportation, Grover's algorithm,
+dense coding, etc.
+
+Other examples of particular quantum systems that are implemented in SymPy
+include second quantization, the simple harmonic oscillator (position/momentum
+and raising/lowering forms) and continuous position/momentum based systems.
```
Comment (flacjacket) [inline quantum_mechanics.tex]: Also particle-in-a-box

File: quantum_mechanics.tex [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+
```
Comment (flacjacket) [inline quantum_mechanics.tex]: Maybe make the above and below a single paragraph?

**Comment by ellisonbg** [comment]
OK I have rebased this on master and done a pretty good pass at the quantum content. Should we merge this and keep iterating in master?

**Comment by certik** [comment]
Thanks @ellisonbg for your work. Thanks for rebasing this on top of @rpmuller's text. I think this looks good overall. It removes @rpmuller's simpler example and adds a more advanced example, which looks good to me.

So overall I think this is good enough, and I am merging this.

**Comment by certik** [comment]
We can now further polish it by sending PRs against master.

**Pull request merged**